content/v2.0/reference/storage-engine.md

```diff
@@ -0,0 +1,174 @@
+Points in a batch do not have to be from the same measurement or tagset.
+
+## Write Ahead Log (WAL)
+
+To ensure durability, we use a Write Ahead Log (WAL).
+<!-- The WAL is a write-optimized storage format that allows for writes to be durable, but not easily queryable -->
+<!-- On the write side, -->
+WAL is a data structure and algorithm that is super simple and powerful.
+It ensures that written data does not disappear when storage engine restarts.
+When a client sends a write request, the following occurs:
+
+1. Write request is appended to the end of the WAL file.
+2. fsync() the data to the file.
+3. Update the in-memory database.
+   <!-- 3. Update the in-memory CACHE? -->
+4. Return success to caller.
+
+fsync() is a system call, so it has a kernel context switch which costs something.
+fsync() takes the file and pushes pending writes all the way through any buffers and caches to disk.
+fsync() is expensive _in terms of time_ but guarantees your data is safe on disk.
+**Important** to batch your points (send in ~2000 points at a time), to fsync() less frequently.
+
+<!-- On read side: -->
+When the storage engine restarts,
+<!-- (if we've pulled the plug, say) -->
+open WAL file and read it back into the in-memory database.
+Answer requests to the /read endpoint.
+
+<!-- ===== V1 material -->
+<!-- TODO is this still true? -->
+<!-- On the file system, the WAL is made up of sequentially numbered files (`_000001.wal`). -->
+<!-- The file numbers are monotonically increasing and referred to as WAL segments. -->
+<!-- When a segment reaches 10MB in size, it is closed and a new one is opened. Each WAL segment stores multiple compressed blocks of writes and deletes. -->
+<!-- Each entry in the WAL follows a [TLV standard](https://en.wikipedia.org/wiki/Type-length-value) with a single byte representing the type of entry (write or delete), a 4 byte `uint32` for the length of the compressed block, and then the compressed block. -->
+
+{{% note%}}
+Once you receive a response to a write request, your data is on disk!
+{{% /note %}}
+
+## Cache
+
+Queries to the storage engine will merge data from the Cache with data from the TSM files.
+Queries execute on a copy of the data that is made from the cache at query processing time.
+This way writes that come in while a query is running won’t affect the result.
+
+The Cache is an in-memory copy of all data points current stored in the WAL.
+The points are organized by the key, which is the measurement, tag set, and unique field.
+Each field is kept as its own time-ordered range.
+The Cache data is not compressed while in memory.
+The cache is recreated on restart by re-reading the WAL files on disk back into memory.
+Deletes sent to the Cache will clear out the given key or the specific time range for the given key.
+
+<!-- It is queried at runtime and merged with the data stored in TSM files. -->
+
+## Time-Structured Merge Tree (TSM)
+
+In order to efficiently compact and store data,
```

Comment:
To efficiently..., the storage engine groups field values by [series](v2.0/reference/key-concepts/data-elements/#series), and then orders...